fix: FIT-31: Ensure media buffers are syncable events that keep audio and video in sync

bmartel opened the pull request:
Previously we were not syncing when a media source would buffer, which led to out of sync audio and video in many cases, especially in spotty network conditions or with very large file sizes. This addresses that issue by syncing buffering state events into SyncManager, and handling the state in the particular cases Audio, Video, Paragraphs. Also updated and aligned the buffering visuals by moving that into the TimelineControls.

_This video showcases a ~300MB video throttled to 3G then seeked and played, and later unthrottled to simulate what was previously causing media sources to go out of sync with one another:_

[Kapture 2025-05-28 at 08.47.55.webm](https://github.com/user-attachments/assets/d3eb401c-ae31-4708-850a-ef52dad397f2)

### Feature Flag:
* Added a new feature flag `FF_SYNCED_BUFFERING` to enable synchronized media buffering across components. (`web/libs/core/src/lib/utils/feature-flags/flags.ts`)

### UI Updates:
* Introduced a visual buffering indicator in the `Timeline` component's controls, including new SCSS styles and animations. (`web/libs/editor/src/components/Timeline/Controls.scss`) [[1]](diffhunk://#diff-896cad12e8256e07c791e6450019f2e6204bdbfac4da1c603c6a0228d4989ed9R2-R27) [[2]](diffhunk://#diff-896cad12e8256e07c791e6450019f2e6204bdbfac4da1c603c6a0228d4989ed9R136-R145)
* Updated the `VideoCanvas` component to conditionally display a buffering indicator based on the buffering state. (`web/libs/editor/src/components/VideoCanvas/VideoCanvas.scss`, `web/libs/editor/src/components/VideoCanvas/VideoCanvas.tsx`) [[1]](diffhunk://#diff-a896199441e4695301ee9d9b080df73cdb2862e219279e3361582472fbe7b84fL42-R42) [[2]](diffhunk://#diff-3d2e7c6f43a23fa8f3ab4f8264e0801e86ce212d3d52a8f85706e72eb6b6a757L569-R613)

### Synchronized Buffering Logic:
* Added a `useSyncedBuffering` hook to manage synchronized buffering states across components using Jotai atoms. (`web/libs/editor/src/hooks/useSyncedBuffering.ts`)
* Updated the `VideoCanvas` and `Html5Player` components to utilize the synchronized buffering logic, including debounced state updates and conditional behavior based on the feature flag. (`web/libs/editor/src/components/VideoCanvas/VideoCanvas.tsx`, `web/libs/editor/src/lib/AudioUltra/Controls/Html5Player.ts`) [[1]](diffhunk://#diff-3d2e7c6f43a23fa8f3ab4f8264e0801e86ce212d3d52a8f85706e72eb6b6a757L109-R123) [[2]](diffhunk://#diff-86a39e502195f50a5f1da05c013b6fd11e320b5c4e75cfed81db83e6ba1de9f7L124-R126)

### Playback Logic Enhancements:
* Modified the `Player` class to include a `buffering` property and handle buffering events (`canplay` and `waiting`) with debounced updates. (`web/libs/editor/src/lib/AudioUltra/Controls/Player.ts`) [[1]](diffhunk://#diff-e870cd503ade4cd0a69eb48059bc9648bf26e175cdfa25fa6b6bbc05cbf8daeeR23) [[2]](diffhunk://#diff-e870cd503ade4cd0a69eb48059bc9648bf26e175cdfa25fa6b6bbc05cbf8daeeR142)
* Adjusted the `Timeline` and `VideoCanvas` components to propagate buffering states through their props and interfaces. (`web/libs/editor/src/components/Timeline/Types.ts`, `web/libs/editor/src/components/Timeline/Timeline.tsx`) [[1]](diffhunk://#diff-92cf542126f556c3070de22992593fbae5a77f0dea843a49252de5df6231a326R18) [[2]](diffhunk://#diff-d4bff87ced2072a8362384fb79468b4f92dffc3aa0d908ae0b4a335d96ac8279R130)

These changes collectively improve the user experience by providing better visual feedback during buffering and ensuring consistent playback behavior across components.

bmartel committed:
fix: FIT-31: Ensure media buffers are syncable events that keep audio and video in sync

bmartel committed:
put all changes behind FF

bmartel committed:
fixing FF usage, aligning buffer styling with all media types, fix control spacing on video

bmartel committed:
tweaking the media buffering animations

bmartel committed:
tweaking the media buffering animations

bmartel committed:
update Audio loader and error colors to use design tokens

bmartel committed:
debounce updates to smooth out the handling of buffering

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-31

bmartel commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15301787592)

bmartel committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15301787592

bmartel committed:
Merge branch 'develop' into 'fb-fit-31'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15301787592

bmartel committed:
video should not progress until buffering is over

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-31

bmartel committed:
use synced buffering

bmartel committed:
don't over report buffering, only report when all buffering elements would change

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-31

bmartel committed:
cleanup and waveform buffer sync to player instance

bmartel committed:
fix syncing A/V + Paragraphs

bmartel committed:
don't allow video to seek while buffering it causes a loop

bmartel committed:
linting

bmartel committed:
debounce the audio player buffering updates so it does not preemptively report to the sync group and cause over buffering

bmartel committed:
linting

bmartel committed:
ff all the audio synced buffering changes

bmartel committed:
linting

bmartel commented on web/libs/editor/src/tags/object/Paragraphs/model.js:
```diff
@@ -292,6 +310,14 @@ const PlayableAndSyncable = types
       self.audioDuration = audio.duration;
     },
 
+    handleCanPlay() {
+      self.handleBuffering(false);
+    },
+
+    handleWaiting() {
+      self.handleBuffering(true);
+    },
+
```
```suggestion
```

bmartel commented on web/libs/editor/src/tags/object/Paragraphs/model.js:
```diff
@@ -257,6 +261,20 @@ const PlayableAndSyncable = types
+      self.buffering = isBuffering;
+      self.trackPlayingId();
+    },
+
+    handleBuffering(isBuffering) {
+      if (!isSyncedBuffering) return;
+      self.buffering = isBuffering;
+      self.triggerSync("buffering", { isBuffering });
```
```suggestion
```

bmartel committed:
Apply suggestions from code review

bmartel committed:
linting

hlomzik commented on web/libs/editor/src/components/Timeline/Controls.scss:
```diff
@@ -110,3 +133,13 @@
+  }
+
+  100% {
+    left: 200%;
```
Animation feels stuck when it fills the whole width. What do you think about making it faster and continuous? I believe changing 200% to 100% will do it.

bmartel committed:
Merge remote-tracking branch 'origin/develop' into fb-fit-31

bmartel committed:
improve animation of buffering loader

bmartel committed:
Merge branch 'develop' into fb-fit-31

Gondragos committed:
fix: replace buffering hook with unified syncable mixin logic

Eliminated the use of `useSyncedBuffering` and transitioned to a direct buffering handling mechanism. This reduces complexity and improves predictability.

- Added `useRefCallback` and `useValueRef` hooks for better callback and state tracking.
- Introduced play/pause method patching to ensure proper async queueing of media operations.
- Updated relevant components for the new buffering management logic.
- fix Player's buffering event delay

Gondragos committed:
Merge branch 'develop' into fb-fit-31

# Conflicts:
#	web/libs/editor/src/tags/object/AudioUltra/view.tsx

Gondragos committed:
Lint fixes

Gondragos committed:
fix: improve buffering logic and syncing

Revised the buffering handling mechanism across components for consistent operation. Introduced a unified `updateBuffering` function and optimized buffering event delays.

- Added timeout-based updates to ensure real buffering state consistency.
- Enhanced `Syncable` mixin to skip locking for specific events during buffering.

Gondragos committed:
fix: add paragraphs buffering logic

Refined paragraphs audio buffering and syncing logic to improve functionality and reduce inconsistencies.

- Added `handleBuffering` and `triggerSyncBuffering` for precise buffering states.
- Introduced patched `play`/`pause` methods on audio in paragraphs for avoiding errors.
- Updated `ParagraphAudio` to handle buffering events dynamically.

Gondragos committed:
fix: re-enable flaky visibility toggle test

(it was fixed before but still was skipped)

Gondragos committed:
fix: unify buffering logic with reusable hook

Replaced buffering logic across components with a reusable `useUpdateBuffering` hook for consistency and maintainability.

- Introduced `useUpdateBuffering` to manage buffering states based in idle state of media elemnet.
- Updated `ParagraphAudio` and `VideoCanvas` to utilize the new hook.

Gondragos committed:
fix: update sync pause handling in paragraphs

Refactored the `pause` sync handler to improve buffering and playback consistency in paragraphs.

- Replaced `stopNow` with a new `handleSyncPause` method to decouple logic.
- Ensured proper handling of buffering states.
- Adjusted `triggerSync` calls to include intended playing state for better synchronization.

Gondragos committed:
Linting

Gondragos committed:
fix: handle unhandled promise rejections in play/pause

Updated `patchPlayPauseMethods` to catch promise rejections and prevent unhandled errors during media playback operations instead of queueing which causes unwanted delays.

Gondragos committed:
Merge branch 'develop' into fb-fit-31

Gondragos committed:
fix: Apply isSyncedBuffering flag and refactoring

Gondragos committed:
fix: update isBuffering logic with flag check

Gondragos committed:
Merge branch 'refs/heads/develop' into fb-fit-31

# Conflicts:
#	web/libs/editor/src/components/Timeline/Controls.tsx
#	web/libs/editor/src/tags/object/Audio/view.tsx
#	web/libs/editor/src/tags/object/Video/HtxVideo.jsx

Gondragos committed:
test: Add tests and helpers

Introduced utilities for network throttling, delay testing, media playback, and synchronization across components.

- Add tests for buffering while interacting with tags containing media elements.
- Added `Network` class to simulate network conditions such as throttling, delay, and cache control.
- Introduced `MediaMixin` and associated helpers for media buffering, playback, and state checking.
- Refactored `VideoView` and `AudioView` to integrate with `MediaMixin` for consistent media handling and improved testing capabilities.
- Add `SyncGroup` helper to process a bunch of `MediaView`s

Gondragos committed:
Tests based small fixes

Gondragos committed:
Cleaning

Gondragos committed:
Merge branch 'refs/heads/develop' into fb-fit-31

# Conflicts:
#	web/libs/editor/src/tags/object/Audio/view.tsx

Gondragos committed:
fix: Fix applying ff

Gondragos committed:
Merge branch 'develop' into fb-fit-31

# Conflicts:
#	web/libs/frontend-test/src/helpers/LSF/AudioView.ts

Gondragos committed:
test: Switch on buffering ff in all tests

Gondragos committed:
test: Add buffering playback sync tests

Added integration tests for media buffering and synchronization during playback. Introduced helper enhancements to support media state validations. Updated `SyncGroup` to allow retry logic for synchronization checks.

Gondragos committed:
fix: Update media syncing during buffering

Updated logic in `Paragraphs`, `Audio`, and `Video` models for better media synchronization while buffering. Introduced `isPaused` state management, `skip_seek_event` flag, and enhanced seek/seek-buffering handling.

Gondragos committed:
Fix tests

Gondragos committed:
fix: Handle browser closure errors in error collector

Gondragos commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16267970518)
> Successfully merged:  create mode 100644 web/libs/editor/src/components/TaskSummary/__tests__/TaskSummary.test.tsx

Gondragos committed:
Merge branch 'develop' into 'fb-fit-31'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16267970518

Gondragos committed:
Fix phrases playing display, optimize rendering, increase stability

hlomzik commented on web/libs/editor/src/components/Timeline/Controls.scss:
```diff
@@ -111,3 +134,13 @@
+  100% {
+    left: 100%;
+  }
+}
```
```suggestion
}

```

Gondragos committed:
fix: Simplify draw handling in Visualizer

Gondragos committed:
Merge branch 'develop' into fb-fit-31

# Conflicts:
#	web/libs/editor/src/components/VideoCanvas/VideoCanvas.tsx
#	web/libs/editor/src/tags/object/Video/HtxVideo.jsx
#	web/libs/frontend-test/src/helpers/LSF/VideoView.ts

Commit:
Accept suggestions

Co-authored-by: Andrew <hlomzik@gmail.com>

robot-ci-heartex committed:
ci: Build tag docs

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16600887656

Gondragos committed:
Merge branch 'develop' into fb-fit-31

# Conflicts:
#	web/libs/frontend-test/src/helpers/LSF/VideoView.ts

Gondragos committed:
Merge branch 'develop' into fb-fit-31

# Conflicts:
#	web/libs/editor/src/tags/object/Paragraphs/HtxParagraphs.jsx

niklub commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16747575052)
> Successfully merged:  delete mode 100644 web/libs/app-common/src/blocks/StorageProviderForm/providers/localFiles.ts

niklub committed:
Merge branch 'develop' into 'fb-fit-31'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16747575052

Commit:
Merge branch 'develop' into fb-fit-31

niklub commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16861459102)
> Successfully merged:  create mode 100644 label_studio/tests/data_import/test_streaming_import.py

niklub committed:
Merge branch 'develop' into 'fb-fit-31'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16861459102

niklub commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16883035943)
> Successfully merged:  create mode 100644 web/libs/editor/src/lib/AudioUltra/Visual/Renderer/ResizeRenderer.ts

niklub committed:
Merge branch 'develop' into 'fb-fit-31'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16883035943

niklub commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16963704331)
> Successfully merged:  delete mode 100644 web/libs/editor/src/regions/ImageRegion.js

niklub committed:
Merge branch 'develop' into 'fb-fit-31'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16963704331

Gondragos commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17040659999)
> Successfully merged:  79 files changed, 1663 insertions(+), 347 deletions(-)

Gondragos committed:
Merge branch 'develop' into 'fb-fit-31'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17040659999

Gondragos committed:
fix: Make Sync buffering playback test more stable

Gondragos committed:
Fix types

Gondragos committed:
Fix pausing before load

Fix for: Failed to set the 'currentTime' property on 'HTMLMediaElement': The provided double value is non-finite

Gondragos commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17044174008)
> Successfully merged:  7 files changed, 159 insertions(+), 20 deletions(-)

Gondragos committed:
Merge branch 'develop' into 'fb-fit-31'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17044174008

Gondragos committed:
Fix auto merge consequences
